app/controllers/api/v1/shared_accesses_controller.rb

```diff
@@ -0,0 +1,78 @@
+# frozen_string_literal: true
+
+module Api
+  module V1
+    class SharedAccessesController < ApplicationController
+      skip_before_action :verify_authenticity_token
+      before_action :find_room
+
+      # POST /api/v1/shared_accesses/room/friendly_id.json
+      def create
+        users = User.where(id: params[:users])
+
+        users.each do |user|
+          SharedAccess.find_or_create_by!(user_id: user.id, room_id: @room.id) if user.room_shareable?(@room)
+        end
+
+        render_json status: :ok
+      end
+
+      # DELETE /api/v1/shared_accesses/room/friendly_id.json
+      def destroy
+        user = User.find_by(id: params[:user_id])
+
+        SharedAccess.find_by!(user_id: user.id, room_id: @room.id).delete
+
+        render_json status: :ok
+      end
+
+      # GET /api/v1/shared_accesses/room/friendly_id/shared_users.json
+      def shared_users
+        shared_users = []
+
+        # User is added to the shared_user list if the room is shared to the user and it is not already included in shared_user
+        User.all.each do |user|
+          shared_users << user if user.room_shared?(@room) && shared_users.exclude?(user)
+        end
+
+        shared_users.map! do |user|
+          {
+            id: user.id,
+            name: user.name,
+            email: user.email,
+            avatar: user.user_avatar
+          }
+        end
+
+        render_json data: shared_users, status: :ok
+      end
+
+      # GET /api/v1/shared_accesses/room/friendly_id/shareable_users.json
+      def shareable_users
+        shareable_users = []
+
+        # User is added to the shareable_user list unless it's the room owner or the room is already shared to the user
+        User.all.each do |user|
+          shareable_users << user if user.room_shareable?(@room)
+        end
```

Comment:
Performance wise - this isn't that good either. We'd probably want to do a query here instead. Get all users where the user's id is not either the room owner's id, or an id of anyone with who it's already shared with

Reply:
```
User.where.not(id: [room_owner_id, list of shared users]
```